engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/generic/DelayedFlinkKafkaConsumer.scala

```diff
@@ -0,0 +1,120 @@
+      runtimeContext.getProcessingTimeService,
+      runtimeContext.getExecutionConfig.getAutoWatermarkInterval,
+      runtimeContext.getUserCodeClassLoader,
+      runtimeContext.getTaskNameWithSubtasks,
+      runtimeContext.getMetricGroup,
+      consumerMetricGroup,
+      deserializer,
+      properties,
+      pollTimeout,
+      useMetrics,
+      timeExtract,
+      delay
+    )
+  }
+}
+
+object DelayedKafkaFetcher {
+  private val maxSleepTime = time.Duration.of(30, ChronoUnit.SECONDS).toMillis
+}
+
+class DelayedKafkaFetcher[T](sourceContext: SourceFunction.SourceContext[T],
+                             assignedPartitionsWithInitialOffsets: util.Map[KafkaTopicPartition, lang.Long],
+                             watermarkStrategy: SerializedValue[WatermarkStrategy[T]],
+                             processingTimeProvider: ProcessingTimeService,
+                             autoWatermarkInterval: lang.Long,
+                             userCodeClassLoader: ClassLoader,
+                             taskNameWithSubtasks: String,
+                             metricGroup: MetricGroup,
+                             consumerMetricGroup: MetricGroup,
+                             deserializer: KafkaDeserializationSchema[T],
+                             kafkaProperties: Properties,
+                             pollTimeout: lang.Long,
+                             useMetrics: Boolean,
+                             timeExtract: (T, Long) => Long,
+                             delay: Long) extends KafkaFetcher[T](sourceContext, assignedPartitionsWithInitialOffsets, watermarkStrategy,
+  processingTimeProvider, autoWatermarkInterval, userCodeClassLoader, taskNameWithSubtasks, deserializer, kafkaProperties, pollTimeout, metricGroup, consumerMetricGroup, useMetrics) with LazyLogging {
+  import DelayedKafkaFetcher._
+
+  override def emitRecordsWithTimestamps(records: util.Queue[T],
+                                         partitionState: KafkaTopicPartitionState[T, TopicPartition],
+                                         offset: Long,
+                                         kafkaEventTimestamp: Long): Unit = {
+    var maxEventTimestamp = 0L
+    records.forEach(new Consumer[T]{
+      override def accept(r: T): Unit = {
+        val recordTimestamp = timeExtract(r, kafkaEventTimestamp)
+        if(recordTimestamp > maxEventTimestamp){
+          maxEventTimestamp = recordTimestamp
+        }
+      }
+    })
+
+    var latency = processingTimeProvider.getCurrentProcessingTime - maxEventTimestamp
+
+    while (delay > latency) {
+      val eventDelay = delay - latency
+      val sleepTime = Math.min(maxSleepTime, eventDelay)
+
+      val logMessage = s"Sleeping for $sleepTime ms of total $eventDelay ms for ${records.size()} events. Max event timestamp is $maxEventTimestamp, fetcher delay is $delay."
+
+      if (sleepTime < maxSleepTime) {
+        logger.debug(logMessage)
+      } else {
+        logger.info(logMessage)
```

Comment:
Why in case when sleepTime exceed maxSleepTime we only log it as an info and sleep nonetheless? Shouldn't be at least some todo/comment here? or maybe we should sleep only maxSleepTime?